src/storage/src/hummock/file_cache/recent_filter.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{BTreeSet, VecDeque};
+use std::fmt::Debug;
+use std::time::{Duration, Instant};
+
+use parking_lot::RwLock;
+
+#[derive(Debug)]
+pub struct RecentFilter<K>
+where
+    K: Eq + Ord + Send + Sync + Debug + Clone + 'static,
+{
+    refresh_interval: Duration,
+    inner: RwLock<CacheRefillFilterInner<K>>,
+}
+
+#[derive(Debug)]
+struct CacheRefillFilterInner<K>
+where
+    K: Eq + Ord + Send + Sync + Debug + Clone + 'static,
+{
+    last_refresh: Instant,
+    layers: VecDeque<RwLock<BTreeSet<K>>>,
+}
+
+impl<K> RecentFilter<K>
+where
+    K: Eq + Ord + Send + Sync + Debug + Clone + 'static,
+{
+    pub fn new(layers: usize, refresh_interval: Duration) -> Self {
+        assert!(layers > 0);
+        let layers = (0..layers)
+            .map(|_| BTreeSet::new())
+            .map(RwLock::new)
+            .collect();
+        let inner = CacheRefillFilterInner {
+            last_refresh: Instant::now(),
+            layers,
+        };
+        let inner = RwLock::new(inner);
+        Self {
+            refresh_interval,
+            inner,
+        }
+    }
+
+    pub fn insert(&self, key: K) {
+        if let Some(mut inner) = self.inner.try_write() {
```

Comment:
IMO, we might need to limit the memory of `RecentFilter` (another PR is ok)

Reply:
I don't think it will be too large for it only counts deduplicated sst id (64bit) inserted during the last minute. A limitation is okay for me if it really goes too large.